### PR TITLE
[master < T565-FL] Use indexes in optional match and foreach

### PR DESCRIPTION
**Pull request by Josipmrden**
Indexes were not applied when using optional match with label property index.The actual problem was that in the additional branch, optional match would not see the bounded symbol when filtering by that symbol's property. Index could not be fetched accordingly

FOREACH. Index was not applied correctly since the plan was forbidden to alter the inner children of the foreach logical operator. Merge can be one of the children, hence the index lookup rewriter can create the index.

[master < Task] PR
- [x] Check, and update documentation if necessary
- [ ] Update [changelog](https://docs.memgraph.com/memgraph/changelog)
- [ ] Provide the full content or a guide for the final git message

### CODE REVIEW

File: src/query/plan/rewrite/index_lookup.hpp [hunk 1]
```diff
@@ -444,7 +444,7 @@ class IndexLookupRewriter final : public HierarchicalLogicalOperatorVisitor {
 
   bool PreVisit(Foreach &op) override {
     prev_ops_.push_back(&op);
-    return false;
+    return true;
```
Comment (antaljanosbenjamin) [inline src/query/plan/rewrite/index_lookup.hpp:447]: This won't visit the internal brancch of the `FOREACH` operator, right?

Comment (Josipmrden) [reply src/query/plan/rewrite/index_lookup.hpp]: it will, my mistake. will correct to rewrite the branch with the new root.

**Review by antoniofilipovic — COMMENTED** [review]
I checked the optional match and it is fixed, and for `foreach` clause I am not sure what was bug, so we can discuss and I will then check that part.

File: tests/unit/query_plan.cpp [hunk 2]
```diff
@@ -662,6 +662,30 @@ TYPED_TEST(TestPlanner, MatchOptionalMatchWhereReturn) {
   DeleteListContent(&optional);
 }
 
+TYPED_TEST(TestPlanner, MatchOptionalMatchNodePropertyWithIndex) {
+  // Test MATCH (n) OPTIONAL MATCH (m) WHERE n.prop = m.prop RETURN n
```
Comment (antaljanosbenjamin) [inline tests/unit/query_plan.cpp:666]: Both `n` and `m` should have label in the query, aren't they? You implemented the right thing, but this comment is a bit incomplete.

File: tests/unit/query_plan.cpp [hunk 3]
```diff
@@ -1685,5 +1709,29 @@ TYPED_TEST(TestPlanner, Foreach) {
         QUERY(SINGLE_QUERY(FOREACH(i, {CREATE(PATTERN(NODE("n")))}), FOREACH(j, {CREATE(PATTERN(NODE("n")))})));
     CheckPlan<TypeParam>(query, storage, ExpectForeach(input, updates), ExpectEmptyResult());
   }
+
+  {
+    // FOREACH with index
```
Comment (antaljanosbenjamin) [inline tests/unit/query_plan.cpp:1714]: Here we just verify that the planner does visit the inside of the `FOREACH` clause right? And I assume (based on the changes I saw) that didn't work before, right?

**Review by antaljanosbenjamin — APPROVED** [review]
I tihnk the changes are fine, only one comment should be extended. Please make a note somewhere about what is the expected behavior of `FOREACH` after the changes. Nothing serious, just something like:
> The inside branch of `FOREACH` will be planned prpoerly using the information inside it, thus it means proper index usage is expected among other changes from the plan rewriter.

A question that I don't know how to answer: does the clauses before the `FOREACH` can affect the planning of the inside branch of `FOREACH`? The reasons I am not sure how to answer it:
1. I don't know exactly how the `RewriteBranch` function works, whether it handles the informations properly. Based on a short look I had I think the filters are not transfered, so I think they won't have affect on the planning of the inside branch.
2. Actually I am also not sure whether we can write something that actually takes into account any filter, because we only allow "write" clauses, such as `FOREACH`, `CREATE`, `DELETE`, `DETACH`, `MERGE`, `REMOVE`, `SET`.

File: tests/unit/query_plan.cpp (hunk 2 above)
Comment (Josipmrden) [reply tests/unit/query_plan.cpp:666]: yes that's true, my mistake

File: tests/unit/query_plan.cpp (hunk 3 above)
Comment (Josipmrden) [reply tests/unit/query_plan.cpp:1714]: yes. It used the ScanAll index rather than indexed scan.

**Review by antoniofilipovic — APPROVED** [review]

**Pull request merged**